File: java/yass/main/ch/softappeal/yass/util/Dumper.java
package ch.softappeal.yass.util;

import java.lang.reflect.Array;
import java.lang.reflect.Field;
import java.util.Collection;
import java.util.IdentityHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class Dumper {

    private static final Set<Class<?>> PRIMITIVE_WRAPPER_CLASSES = Set.of(
        Boolean.class, Byte.class, Short.class, Integer.class, Long.class, Float.class, Double.class
    );

    private final boolean compact;
    private final boolean referenceables;

    private final Set<Class<?>> concreteValueClasses;
    protected final boolean isConcreteValueClass(final Class<?> type) {
        return concreteValueClasses.contains(type);
    }

    /**
     * @param compact one-liner or multiple lines
     * @param referenceables true: dumps graphs (objects are marked with #); false: dumps trees
     * @param concreteValueClasses only allowed if (referenceables); these objects should not reference others; do not print # for these classes
     */
    public Dumper(final boolean compact, final boolean referenceables, final Class<?>... concreteValueClasses) {
        this.compact = compact;
        this.referenceables = referenceables;
        if (!referenceables && (concreteValueClasses.length != 0)) {
            throw new IllegalArgumentException("concreteValueClasses only allowed if (referenceables)");
        }
        this.concreteValueClasses = Set.of(concreteValueClasses);
    }

    public final StringBuilder append(final StringBuilder s, final @Nullable Object value) {
        new Dump(s).dump(value);
        return s;
    }

    public final String dump(final @Nullable Object value) {
        return append(new StringBuilder(256), value).toString();
    }

    private final Map<Class<?>, List<Field>> class2fields = new ConcurrentHashMap<>();

    private final class Dump {
        private final StringBuilder out;
        Dump(final StringBuilder out) {
            this.out = Objects.requireNonNull(out);
        }
        private int tabs = 0;
        private void appendTabs() {
            for (var t = tabs; t > 0; t--) {
                out.append("    ");
            }
        }
        private void appendLine() {
            out.append('\n');
        }
        private void inc(final String s) {
            out.append(s);
            appendLine();
            tabs++;
        }
        private void dec(final String s) {
            tabs--;
            appendTabs();
            out.append(s);
        }
        private void dumpArray(final Object array) {
            final var length = Array.getLength(array);
            if (compact) {
                out.append("[ ");
                for (var i = 0; i < length; i++) {
                    dump(Array.get(array, i));
                    out.append(' ');
                }
                out.append(']');
            } else {
                inc("[");
                for (var i = 0; i < length; i++) {
                    appendTabs();
                    dump(Array.get(array, i));
                    appendLine();
                }
                dec("]");
            }
        }
        private void dumpCollection(final Collection<?> collection) {
            if (compact) {
                out.append("[ ");
                for (final Object e : collection) {
                    dump(e);
                    out.append(' ');
                }
                out.append(']');
            } else {
                inc("[");
                for (final Object e : collection) {
                    appendTabs();
                    dump(e);
                    appendLine();
                }
                dec("]");
            }
        }
        private void dumpMap(final Map<?, ?> map) {
            if (compact) {
                out.append("{ ");
                for (final Map.Entry<?, ?> entry : map.entrySet()) {
                    dump(entry.getKey());
                    out.append("->");
                    dump(entry.getValue());
                    out.append(' ');
                }
                out.append('}');
            } else {
                inc("{");
                for (final Map.Entry<?, ?> entry : map.entrySet()) {
                    appendTabs();
                    dump(entry.getKey());
                    out.append(" -> ");
                    dump(entry.getValue());
                    appendLine();
                }
                dec("}");
            }
        }
        private void dumpClassFields(final List<Field> fields, final Object object) {
            for (final var field : fields) {
                final @Nullable Object value;
                try {
                    value = field.get(object);
                } catch (final IllegalAccessException e) {
                    throw new RuntimeException(e);
                }
                if (value != null) {
                    if (compact) {
                        out.append(field.getName()).append('=');
                        dump(value);
                        out.append(' ');
                    } else {
                        appendTabs();
                        out.append(field.getName()).append(" = ");
                        dump(value);
                        appendLine();
                    }
                }
            }
        }
        private final @Nullable Map<Object, Integer> alreadyDumpedObjects = referenceables ? new IdentityHashMap<>(16) : null;
        private void dumpClass(final Class<?> type, final Object object) {
            final var referenceables = Dumper.this.referenceables && !isConcreteValueClass(type);
            final int index;
            if (referenceables) {
                final var reference = alreadyDumpedObjects.get(object);
                if (reference != null) {
                    out.append('#').append(reference);
                    return;
                }
                index = alreadyDumpedObjects.size();
                alreadyDumpedObjects.put(object, index);
            } else {
                index = 0;
            }
            if (!Dumper.this.dumpValueClass(out, type, object)) {
                final var fields = class2fields.computeIfAbsent(type, Reflect::allFields);
                if (compact) {
                    out.append(type.getSimpleName()).append("( ");
                    dumpClassFields(fields, object);
                    out.append(')');
                } else {
                    inc(type.getSimpleName() + '(');
                    dumpClassFields(fields, object);
                    dec(")");
                }
            }
            if (referenceables) {
                out.append('#').append(index);
            }
        }
        void dump(final @Nullable Object value) {
            if (value == null) {
                out.append("null");
            } else if (value instanceof CharSequence) {
                out.append('"').append(value).append('"');
            } else if (value instanceof Collection) {
                dumpCollection((Collection<?>)value);
            } else if (value instanceof Map) {
                dumpMap((Map<?, ?>)value);
            } else {
                final var type = value.getClass();
                if (type.isEnum() || PRIMITIVE_WRAPPER_CLASSES.contains(type)) {
                    out.append(value);
                } else if (type.isArray()) {
                    dumpArray(value);
                } else if (type == Character.class) {
                    out.append('\'').append(value).append('\'');
                } else {
                    dumpClass(type, value);
                }
            }
        }
    }

    /**
     * Could dump a value class (these should not reference other objects). Should be an one-liner.
     * This implementation does nothing and returns false.
     * @return true: if we dumped object of type to out; false: use default implementation
     */
    protected boolean dumpValueClass(final StringBuilder out, final Class<?> type, final Object object) {
        return false;
    }

}

File: kotlin/yass/main/ch/softappeal/yass/Dumper.kt
package ch.softappeal.yass

import java.lang.reflect.Array
import java.lang.reflect.Field
import java.util.IdentityHashMap
import java.util.concurrent.ConcurrentHashMap

/** Dumps a value class (these should not reference other objects) to out (should be an one-liner). */
typealias ValueDumper = (out: StringBuilder, value: Any) -> Unit

val EmptyValueDumper: ValueDumper = { _, _ -> }

typealias Dumper = (out: StringBuilder, value: Any?) -> StringBuilder

@JvmOverloads
fun treeDumper(compact: Boolean, valueDumper: ValueDumper = EmptyValueDumper): Dumper =
    dumper(compact, false, emptySet(), valueDumper)

@JvmOverloads
fun graphDumper(
    compact: Boolean,
    concreteValueClasses: Set<Class<*>> = emptySet(),
    valueDumper: ValueDumper = EmptyValueDumper
): Dumper =
    dumper(compact, true, concreteValueClasses, valueDumper)

fun Dumper.dump(value: Any?): StringBuilder = this(StringBuilder(256), value)

fun Dumper.toString(value: Any?): String = this.dump(value).toString()

private val PrimitiveWrapperClasses = setOf(
    Boolean::class::javaObjectType.get(),
    Byte::class::javaObjectType.get(),
    Short::class::javaObjectType.get(),
    Int::class::javaObjectType.get(),
    Long::class::javaObjectType.get(),
    Float::class::javaObjectType.get(),
    Double::class::javaObjectType.get()
)

private fun dumper(compact: Boolean, graph: Boolean, concreteValueClasses: Set<Class<*>>, valueDumper: ValueDumper): Dumper {
    val class2fields = ConcurrentHashMap<Class<*>, List<Field>>()
    return { out, value ->
        var tabs = 0
        val alreadyDumpedObjects = if (graph) IdentityHashMap<Any, Int>(16) else null
        fun appendTabs() {
            for (t in tabs downTo 1) out.append("    ")
        }

        fun appendLine() {
            out.append('\n')
        }

        fun inc(s: CharSequence) {
            out.append(s)
            appendLine()
            tabs++
        }

        fun dec(s: CharSequence) {
            tabs--
            appendTabs()
            out.append(s)
        }

        fun append(s: CharSequence) {
            out.append(s)
        }

        fun append(c: Char) {
            out.append(c)
        }

        fun dump(value: Any?) {
            fun dumpArray(array: Any) {
                val length = Array.getLength(array)
                if (compact) {
                    append('[')
                    var first = true
                    for (i in 0 until length) {
                        if (!first) append(',')
                        first = false
                        dump(Array.get(array, i))
                    }
                    append(']')
                } else {
                    inc("[")
                    for (i in 0 until length) {
                        appendTabs()
                        dump(Array.get(array, i))
                        appendLine()
                    }
                    dec("]")
                }
            }

            fun dumpCollection(collection: Collection<*>) {
                if (compact) {
                    append('[')
                    var first = true
                    for (e in collection) {
                        if (!first) append(',')
                        first = false
                        dump(e)
                    }
                    append(']')
                } else {
                    inc("[")
                    for (e in collection) {
                        appendTabs()
                        dump(e)
                        appendLine()
                    }
                    dec("]")
                }
            }

            fun dumpMap(map: Map<*, *>) {
                if (compact) {
                    append('{')
                    var first = true
                    for ((k, v) in map) {
                        if (!first) append(',')
                        first = false
                        dump(k)
                        append('>')
                        dump(v)
                    }
                    append('}')
                } else {
                    inc("{")
                    for ((k, v) in map) {
                        appendTabs()
                        dump(k)
                        append(" -> ")
                        dump(v)
                        appendLine()
                    }
                    dec("}")
                }
            }

            fun dumpClass(type: Class<*>, obj: Any) {
                val checkDumped = graph && !concreteValueClasses.contains(type)
                val index: Int
                if (checkDumped) {
                    val reference = alreadyDumpedObjects!![obj]
                    if (reference != null) {
                        append("#$reference")
                        return
                    }
                    index = alreadyDumpedObjects.size
                    alreadyDumpedObjects[obj] = index
                } else
                    index = 0
                val oldLength = out.length
                valueDumper(out, obj)
                if (oldLength == out.length) {
                    val fields = class2fields.computeIfAbsent(type) { allFields(it) }
                    fun dumpFields() {
                        var first = true
                        for (field in fields) {
                            val f = field.get(obj)
                            if (f != null) {
                                if (compact) {
                                    if (!first) append(',')
                                    first = false
                                    append("${field.name}=")
                                    dump(f)
                                } else {
                                    appendTabs()
                                    append("${field.name} = ")
                                    dump(f)
                                    appendLine()
                                }
                            }
                        }
                    }
                    if (compact) {
                        append("${type.simpleName}(")
                        dumpFields()
                        append(')')
                    } else {
                        inc("${type.simpleName}(")
                        dumpFields()
                        dec(")")
                    }
                }
                if (checkDumped) append("#$index")
            }
            when (value) {
                null -> append("null")
                is CharSequence -> append("\"$value\"")
                is Collection<*> -> dumpCollection(value)
                is Map<*, *> -> dumpMap(value)
                else -> {
                    val type = value.javaClass
                    when {
                        type.isEnum || PrimitiveWrapperClasses.contains(type) -> out.append(value)
                        type.isArray -> dumpArray(value)
                        type === Character::class.java -> append("'$value'")
                        else -> dumpClass(type, value)
                    }
                }
            }
        }
        dump(value)
        out
    }
}
